engine/src/main/scala/cromwell/engine/workflow/WorkflowStoreActor.scala

```diff
@@ -0,0 +1,145 @@
+package cromwell.engine.workflow
+
+import java.time.OffsetDateTime
+import java.util.UUID
+
+import akka.actor.{Actor, Props}
+import akka.event.Logging
+import cromwell.core.{WorkflowId, WorkflowSourceFiles}
+import cromwell.database.obj.WorkflowMetadataKeys
+import cromwell.engine.workflow.WorkflowStore.WorkflowToStart
+import cromwell.engine.workflow.WorkflowStoreActor._
+import cromwell.services.MetadataServiceActor.PutMetadataAction
+import cromwell.services.{MetadataEvent, MetadataKey, MetadataValue, ServiceRegistryClient}
+
+import scala.language.postfixOps
+import scala.collection._
+
+object WorkflowStore {
+  sealed trait WorkflowState {
+    def isStartable: Boolean = {
+      this match {
+        case x: StartableState => true
+        case _ => false
+      }
+    }
+  }
+
+  case object Running extends WorkflowState
+
+  sealed trait StartableState extends WorkflowState
+  case object Submitted extends StartableState
+  case object Restartable extends StartableState
+
+  final case class SubmittedWorkflow(sources: WorkflowSourceFiles, state: WorkflowState) {
+    def toWorkflowToStart(id: WorkflowId): WorkflowToStart = {
+      state match {
+        case r: StartableState => WorkflowToStart(id, sources, r)
+        case _ => throw new IllegalArgumentException("This workflow is not currently in a startable state")
+      }
+    }
+  }
+
+  final case class WorkflowToStart(id: WorkflowId, sources: WorkflowSourceFiles, state: StartableState)
+}
+
+/**
+  * Stores any submitted workflows which have not completed.
+  *
+  * This is not thread-safe, although it should not be being used in a multithreaded fashion. If that
+  * changes, revisit.
+  */
+trait WorkflowStore {
+  import WorkflowStore._
+
+  /*
+    The LinkedHashMap preserves insertion order which means that any requests for runnable workflows will
+    be satisfied by taking the oldest available.
+   */
+  private var workflowStore = new mutable.LinkedHashMap[WorkflowId, SubmittedWorkflow]
```

Comment:
I was picturing it being like most of Ye Olde Worlde where there was a DB backing it but in memory handling for moment to moment stuff. But you're right that it might not end up that way.